Problem (Fix #82): reward mechanisms not defined precisely

yihuang opened the pull request:
Solution:
- Add reward mechanisms in terms of abci events and rust code

yihuang commented:
> This documentation should be detailed enough for non-developers to understand how rewards work. I guess, it'd be good to provide an overview of everything in "English" (for non-developers) and also in "Pseudo-code" (for developers).
> 
> Currently, it just copy-pastes the implementation which is as good as reading code.

The abstract part should be enough for non-developers?
For the implementation part, I find the rewarding implementation is quite straight forward, the real code is already readable enough for developers I think.

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,287 @@
+
+- Fixed total supply.
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount a validator received should be proportional to the voting power.
```
the last point is a bit unclear -- the emitted amount to all validators is proportional to the total vote power. The received share should be proportional only to participation?

yihuang replied:
I changed to "The rewards amount distributed to a validator should be proportional to his voting power", the minted amount will be explained in the monetary expansion section, it's related to total staking, but not necessarily proportional.

tomtau commented on docs/modules/rewarding.md:
````diff
@@ -0,0 +1,287 @@
+
+```rust
+pub struct RewardsParameters {
+    /// Time inteval in seconds to do rewards distribution
````
```suggestion
    /// Time interval in seconds to do rewards distribution
```

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,287 @@
+    /// Monetary expansion formula parameter
+    pub monetary_expansion_tau: u64,
+    /// Monetary expansion formula parameter
+    pub monetary_expansion_decay: u64,
```
required ranges?

tomtau replied:
should `monetary_expansion_tau` be called `initial_tau` or something like that if it decays?

yihuang replied:
> required ranges?

Added the validation code.

yihuang replied:
> should `monetary_expansion_tau` be called `initial_tau` or something like that if it decays?

Since they are all in `InitNetworkParameters`, I think the init part if implicit already?

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,287 @@
+    /// Record how many coins have been minted, can't exceed the cap
+    pub minted: Coin,
+    /// Parameter in monetary expansion formula, decayed for each rewards distribution
+    pub tau: u64,
```
`current_tau` to make it clearer that it may be different from the initial one?

yihuang replied:
Done

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,287 @@
+        self.last_state
+            .as_mut()
+            .unwrap()
+            .validators
```
this is a sort of "implementation detail" and likely to change

yihuang replied:
Cleaned up the code, moved the accidental things into separated functions.

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,287 @@
+                if let Some(voting_power) = self
+                    .validator_state_helper
+                    .get_voting_power(staking_address)
+                {
```
same

tomtau replied:
many unnecessary details all below (plus the semantics of many of these detailed items isn't define, unless one looks at the code right now, so can be out of date -- e.g. `self.tendermint_validator_addresses`)

yihuang replied:
Cleaned up the code, moved the accidental things into separated functions.

devashishdxt commented on docs/modules/rewarding.md:
````diff
@@ -0,0 +1,304 @@
+### Network parameters
+
+```rust
+pub struct RewardsParameters {
````
At the start of the document, can you explain the role each of these network parameters and how they change reward behavior?

yihuang replied:
I think the parameters are all explained in different places in part 1.

devashishdxt replied:
Where is part 1? Am I missing something?

yihuang replied:
> Where is part 1? Am I missing something?

I mean the section "Abstract".
For example, the monetary expansion parameters R0/tau are in the monetary expansion formula.
The reward period has a section in concepts. I should add more description for the `rewards_monetary_expansion_cap` and `monetary_expansion_decay `.

yihuang replied:
Done

devashishdxt commented on docs/modules/rewarding.md:
````diff
@@ -0,0 +1,304 @@
+
+Signing voters are defined as:
+
+```rust
````
One needs to know **Rust** and **ABCI internals** to understand what are **Signing voters**. Maybe add a bit more description?

yihuang replied:
"All the active validators who follow the protocol honestly will be recorded as signing voters."
Added this sentence, should be enough for people to understand it.

devashishdxt commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+
+#### Reward period
+
+Rewards are distributed to the validators periodically.
```
Also specify that this can be changed using network parameter in `genesis.json`.

yihuang replied:
Done

devashishdxt commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+
+During the reward period, the following sources of rewarding are accumulated into reward pool:
+
+- Monetary expansion
```
What does monetary expansion mean? Add a small definition and also link to the **Monetary expansion** section.

yihuang replied:
Added a link.

devashishdxt commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+
+### ABCI Events
+
+#### Begin block
```
Maybe specify **_in English_**, what should happen during `BeginBlock` and don't expose the whole implementation (as it may change in future).

yihuang replied:
Added some simple descriptions before that code.

devashishdxt commented on docs/modules/rewarding.md:
````diff
@@ -0,0 +1,304 @@
+    }
+```
+
+### Hooks
````
No explanation of these **Hooks** and also no mention of where these should be called from.

yihuang replied:
Added more descriptions in the hook section, please review again.

devashishdxt commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+P = rewards_config["distribution_period"]
+
+# total bonded amount of the active validators at the end of the reward epoch
+S = total_staking
```
How is `total_staking` calculated?

yihuang replied:
There is a "total staking" section now which explains it.

yihuang commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- Fixed total supply.
+- Fairness, all honest participators should get rewards, non-participators should not.
```
The sentence after it explains it: "all honest participators should get rewards, non-participators should not".

tomtau replied:
where does this definition come from? 
the common one is:
```
3 (fairness). A process gets a reward for a block if and only if it has a reward parameter for that block equal to 1. 
3' (eventual fairness). There exists a height H such that for a block in the blockchain at height H′ > H a process gets a reward for that block if and only if it has a reward parameter for that block equal to 1.
```

yihuang commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+
+Rewarding module enable crypto chain to reward protocol participators, the design should meet following requirements:
+
+- Fixed total supply.
```
Changed it to "The total supply of CRO is fixed"

yihuang commented:
Just cleaned up the code, removed the accidental things into separated functions.
Now it's very little thing that can be removed without affecting the logic.

yihuang commented:
> There are still a few missing pieces which this spec does not explain:
> 
> * What happens when a validator gets jailed in a reward period?
> * What happens when a validator unbonds their funds in a reward period?

They are explained in the section hooks, added more descriptions.

> 
> Also, it is not clear what should happen in each ABCI call `InitChain`, `BeginBlock`, `EndBlock`, etc. One has to read and understand the whole code to know the behavior of rewards module.
> 

The code is cleaned up now, with comments, it should be quite readable to a developer.

> I think it is better to merge this with [Validator Rewards](https://github.com/crypto-com/chain-docs/blob/master/docs/getting-started/staking.md#validator-rewards) section in main documentation (move everything to a new file?) which already explains everything in good detail.

yihuang commented:
> the doc should be clear and compact, so one can easily judge whether the code does what it should do -- if there's a copy-pasted super-detailed code (rather than just high level snippets), it'll likely get out of date... and it may hard to judge whether the evolved code still does what it should be doing

I think with the cleaned-up version of code, there is very little that is not directly related to the logic. And it's the more precise way to describe the logic.

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,334 @@
+                remains = remains.saturating_sub(amount);
+                distribution.push((*addr, amount));
+            }
+            remains
```
this is still "unfair" (as per the agreed definition):

1. it only tracks "IF" the validator correctly participated, not how often
2. it distributes per vote power share, not per participation.

--
(eventually) fair reward distribution should reward a node for each block if and only if it participated correctly (live, non-byzantine).

...
Here, the counter-example is that if there are more than 1 block during that period, it won't be fair. e.g. a node:
block 1 -- participated (gets recorded)
block 2 -- ... no ...
block 3 -- ... no ...
...
blocks N -- distribution -- gets rewarded for all blocks up to N, regardless how it behaved in each of them which is unfair.

yihuang replied:
You seem to try to distribute rewards per block, but I think they are actually equivalent:
```
block 1 -- reward parameter=1, power of A=p, power of others=p'
block 2 -- reward parameter=0
block 3 -- reward parameter=0
...
blocks N -> distribution
``` 
Assuming all the other validators have reward parameter=1, then the reward distribution of A is:

```
total_power = p + p' * N

# current computation
reward_a = reward_pool * p / total_power

# only focus on block 1
reward_block1 = reward_pool * (p + p') / total_power
reward_a = reward_block1 * p / (p + p')
         = reward_pool * p / total_power
```

yihuang replied:
> where does this definition come from?
> the common one is:
> 
> 3 (fairness). A process gets a reward for a block if and only if it has a reward parameter for that block equal to 1. 
> 3' (eventual fairness). There exists a height H such that for a block in the blockchain at height H′ > H a process gets a reward for that block if and only if it has a reward parameter for that block equal to 1.

I think it's just the 3 in layman's term?

tomtau replied:
normal fairness gets that reward for that block.

eventual fairness gets reward for that block in some time in the future -- the important bit here there is " if and only if it ".

In the current mechanism, yes, it gets a reward for that block AND it also gets rewards for blocks where its reward parameter=0 (as the current mechanism doesn't distinguish correct participation) which it shouldn't

yihuang replied:
> normal fairness gets that reward for that block.
> 
> eventual fairness gets reward for that block in some time in the future -- the important bit here there is " if and only if it ".
> 
> In the current mechanism, yes, it gets a reward for that block AND it also gets rewards for blocks where its reward parameter=0 (as the current mechanism doesn't distinguish correct participation) which it shouldn't

As I showed in the previous comment, it only gets the reward for blocks where its reward parameter = 1, not get rewards for blocks where its reward parameter=0.
I copy it here:

> You seem to try to distribute rewards per block, but I think they are actually equivalent:
> ```
> block 1 -- reward parameter=1, power of A=p, power of others=p'
> block 2 -- reward parameter=0
> block 3 -- reward parameter=0
> ...
> blocks N -> distribution
> ``` 
> Assuming all the other validators have reward parameter=1, then the reward distribution of A is:
> 
> ```
> total_power = p + p' * N
> 
> # current computation
> reward_a = reward_pool * p / total_power
> 
> # only focus on block 1
> reward_block1 = reward_pool * (p + p') / total_power
> reward_a = reward_block1 * p / (p + p')
>          = reward_pool * p / total_power
> ```
> 
>

yihuang replied:
Do you mean don't accumulate rewards into a pool for whole period, but calculate reward distribution for each block separately. Because different blocks might have different fees collected?

tomtau replied:
it's defined with "IF AND ONLY IF" -- e.g. two validators A B, both have the same vote power.
Node A:
```
block 1 -- reward parameter=1
block 2 -- reward parameter=0
block 3 -- reward parameter=0
...
```
Node B:
```
block 1 -- reward parameter=1
block 2 -- reward parameter=1
block 3 -- reward parameter=1
...
```

reward gets distributed at N -- currently, the share calculation takes into account only vote power... so here, A and B have the same vote power, so they get reward the same.
If it's interpreted per block, this does not satisfy (eventual) fairness, as participants should be reward IF AND ONLY IF... here, A was rewarded for blocks 2, 3...

yihuang replied:
> it's defined with "IF AND ONLY IF" -- e.g. two validators A B, both have the same vote power.
> Node A:
> 
> ```
> block 1 -- reward parameter=1
> block 2 -- reward parameter=0
> block 3 -- reward parameter=0
> ...
> ```
> 
> Node B:
> 
> ```
> block 1 -- reward parameter=1
> block 2 -- reward parameter=1
> block 3 -- reward parameter=1
> ...
> ```
> 
> reward gets distributed at N -- currently, the share calculation takes into account only vote power... so here, A and B have the same vote power, so they get rewarded the same.
> When interpreted per block, this does not satisfy (eventual) fairness, as participants should be reward IF AND ONLY IF... here, A was rewarded for blocks 2, 3...

In this case, the computation is like this:
```
pa = power of A
pb = power of B
sum_power = pa + pb * N
sum_power_a = pa
sum_power_b = pb * N
reward_a = reward_pool * sum_power_a / sum_power
reward_b = reward_pool * sum_power_b / sum_power
```

tomtau replied:
Oh, https://github.com/crypto-com/chain-docs/pull/94/files#diff-2c819a76ed9cc68e5c8a8d51dc8ac205R208 there's a `saturating_add`

tomtau replied:
I skimmed through it and only saw "insert"

tomtau replied:
Anyway, this shows that a more high-level concise description or pseudocode will be clearer than 100+ lines of detailed code where it is hard to spot the important parts

tomtau replied:
`record_voters_for_rewarding` could just be something like:
```
if signed_voters[staked_address]:
signed_voters[staked_address] += vote_power_at_X
else:
signed_voters[staked_address] = vote_power_at_X
```

yihuang replied:
> `record_voters_for_rewarding` could just be something like:
> 
> ```
> if signed_voters[staked_address]:
> signed_voters[staked_address] += vote_power_at_X
> else:
> signed_voters[staked_address] = vote_power_at_X
> ```

But it will miss two informations:
- overflow behavior
- Error handling when address and voting power not found

tomtau replied:
Is any of them expected to happen as a part of normal behaviour?

It'll ideally list all its assumptions / invariants separately -- e.g. it's assumed that one can lookup staking address (based on the consensus pubkey/address) for the duration of the reward period, unless it's been removed due to punishment mechanisms?

For expected behaviour, it should specify these.
For unexpected behaviour, there should be an overall note around handling -- e.g. unexpected behaviour / invariant breaking behaviour will be logged with "error" level in deployment builds, but will panic in debug/fuzzing/test builds?

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,334 @@
+
+- The total supply of CRO is fixed.
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount distributed to a validator should be proportional to his voting power.
```
The total reward emitted should be proportional to the total voting power,

but the distributed amount should be proportional to the correct participation (assuming distribution doesn't happen every block due to performance degrade).

yihuang replied:
total reward emitted could be positively or negatively related to the total voting power, not necessarily proportional(maybe I have a wrong understanding of proportional?).
"correct participation" is the second point?

tomtau replied:
I'm confused if the third point talks about the share or the total emitted reward?

yihuang replied:
> I'm confused if the third point talks about the share or the total emitted reward?

It says "The rewards amount distributed to a validator", so it's the share of a single validator, how should I make it clearer?

tomtau replied:
Well, one of the points of this was:

1. total emitted amount should be proportional to the total vote power (as before)
2. individual shares should correspond to correct behaviour participation, *regardless* of the vote power

And the correct behaviour is defined with the respect to individual blocks with "IF AND ONLY IF" for each block -- it's not defined as "As long as a validator exhibited correct behaviour in 1 block out of N in a given, he should get reward for all N blocks"

yihuang replied:
“total emitted amount should be proportional to the total vote power” is not accurate, fees and slashings are not related to total vote power, the amount of new minted coins is a function of total vote power, but not necessarily proportional to that.

By "regardless of the vote power", do you mean one vote of A power and one vote of B power should get same amount of reward?

tomtau replied:
> “total emitted amount should be proportional to the total vote power” is not accurate, fees and slashings are not related to total vote power, the amount of new minted coins is a function of total vote power, but not necessarily proportional to that.

emitted means it's newly emitted, so doesn't include fees or slashing -- perhaps more precise is inversely proportional or what's out of the formula.
it can say that the "total reward includes total emitted coins (proportional to ...) and bonus"

> By "regardless of the vote power", do you mean one vote of M power and one vote of N power should get the same amount of reward?
It depends on their correct behaviour.
If both behaved correctly (live, non-byzantine) at all times, yes

yihuang replied:
> > By "regardless of the vote power", do you mean one vote of M power and one vote of N power should get the same amount of reward?
> 
> It depends on their correct behaviour.
> If both behaved correctly (live, non-byzantine) at all times, yes

Do you mean a validator with 10 power get the same reward as a validator with 1 power, if they both behaved correctly at all times? Isn't this just incentivize the validator with more power to separate his power into multiple smaller validators, and takes more validator spots, and which is something we want to avoid?

tomtau replied:
Ignore that -- it was my misunderstanding of how the activity is recorded.

yihuang commented:
> looks a bit simpler, but again the code snippets should ideally omit details that may change or explain them more in detail.
> e.g. `self.rewards_pool_updated` is just some temporary flag (the implementation may change) that helps with app hash re-calculations or what not?

`self.rewards_pool_updated` is part of the logic of how `last_block_height` gets updated.

yihuang commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,304 @@
+- Fairness, all honest participators should get rewards, non-participators should not.
+- The rewards amount a validator received should be proportional to the voting power.
+
+The basic idea is to record signing voters for each block, along with the voting power at the vote time. Distribute the reward pool according to the voting powers.
```
Added an example now.

yihuang commented:
> A few thoughts:
> 
> This document exposes a lot of implementation details (it exposes the whole implementation). Specifications, generally, are supposed to act as a blueprint to help developers implement the functionality. It should describe all the flows in the functionality and also specify behavior of the program for each flow. Besides this, it is ok to provide high level interfaces that can be used by developer but, detailed implementation shouldn't be a part of specification.
> 
> Understanding the behaviour of rewards module from this document requires going through the code and understanding it, which is not what specification is meant for. Ideally, `specification + developer = code`. Whereas here, `specification + code + developer = more code`.
> 
> Any thoughts?

I guess it all come down to how precise do we want the specification to be, or what functionality do we want from the specification, do we want it to be some descriptive documentation, or precise stuff which can be used to check the correctness.
For example, the error handling(panic or not in some edge cases) or overflow behavior(panic, wrap, saturate) is important part of the correctness, many seemingly verboseness in the code are  actually these things.

devashishdxt commented:
> > A few thoughts:
> > This document exposes a lot of implementation details (it exposes the whole implementation). Specifications, generally, are supposed to act as a blueprint to help developers implement the functionality. It should describe all the flows in the functionality and also specify behavior of the program for each flow. Besides this, it is ok to provide high level interfaces that can be used by developer but, detailed implementation shouldn't be a part of specification.
> > Understanding the behaviour of rewards module from this document requires going through the code and understanding it, which is not what specification is meant for. Ideally, `specification + developer = code`. Whereas here, `specification + code + developer = more code`.
> > Any thoughts?
> 
> I guess it all come down to how precise do we want the specification to be, or what functionality do we want from the specification, do we want it to be some descriptive documentation, or precise stuff which can be used to check the correctness.
> For example, the error handling(panic or not in some edge cases) or overflow behavior(panic, wrap, saturate) is important part of the correctness, many seemingly verboseness in the code are actually these things.

These are important parts of a program's implementation and most of the specifications don't delve into these implementation details because the idea to to keep them agnostic of how any developer chooses to implement them (for example, decisions like which data sctucture to use, what language to use, how to handle errors, etc. are left to developers). Here is an example, [this](https://www.ietf.org/rfc/rfc793.txt) is **Transmission Control Protocol** specification and they explain the whole protocol without code.

P.S.: I'm ok with writing the whole code in specification if everyone is comfortable with this.

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,389 @@
+- [Transaction Fees](../getting-started/transaction.md#transaction-fee)
+- Slashings
+
+The reward pool get accumulated through the whole reward period, it get distributed once at the end of reward period.
```
```suggestion
The reward pool gets accumulated throughout the whole reward period; it gets distributed once at the end of each reward period.
```

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,389 @@
+
+The reward pool get accumulated through the whole reward period, it get distributed once at the end of reward period.
+
+Because of integer calculations, there will be a few remained funds after distribution, they should be left to the next period.
```
```suggestion
Due to fixed point calculations, there may be a few remainders from arithmetic operations after distribution. These remainder amounts should be left in the rewards pool for the next period.
```

tomtau commented on docs/modules/rewarding.md:
```diff
@@ -0,0 +1,389 @@
+
+Because of integer calculations, there will be a few remained funds after distribution, they should be left to the next period.
+
+If there is zero participator for the period, the rewards pool should be left to the next period, rather than get burned.
```
does this mean that if there were no participants, reward distribution should not happen? When would this happen?

yihuang replied:
Maybe, all the previous participators get jailed in node state somehow, but still manage to produce the last block of reward period, because tendermint validator set doesn't updated yet.

tomtau commented on docs/modules/rewarding.md:
````diff
@@ -0,0 +1,389 @@
+##### `begin_block`
+
+```rust
+fn begin_block(req: RequestBeginBlock) {
````
just side-note that I guess it should happen after punishment-related begin block processing? there's a separate issue for this / how to generate validator updates: https://github.com/crypto-com/chain-docs/issues/81 so no need in this PR
